### PR TITLE
[USM] move generic usm files to usm package

### PR DESCRIPTION
**Pull request by Yumasi**
### What does this PR do?

This PR is a first step in refactoring USM. Its goal, is to move the core, generic parts of USM to their own `usm` package. Until now, the core of USM was in the `pkg/network/protocols/http` package. USM now covers more than just HTTP, hence this refactor.

The TLS files were also moved to the `usm` package. This is due to the TLS support code being dependent to both `ebpfSubprogram` (which was moved to the `usm` package) and `http`. It had to be moved in the `usm` package to prevent creating a circular dependency between the `http` `usm` packages. Breaking this circular dependency would require a refactor of how we handle those subprograms, which is out of the scope of this PR.

This change is intentionally limited to moving things around, without changing the code too much. This should come in a later PR.

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/tracer/tracer.go [hunk 1]
```diff
@@ -860,11 +861,11 @@ func (t *Tracer) DebugDumpProcessCache(ctx context.Context) (interface{}, error)
 	return nil, nil
 }
 
-func newHTTPMonitor(c *config.Config, tracer connection.Tracer, bpfTelemetry *telemetry.EBPFTelemetry, offsets []manager.ConstantEditor) *http.Monitor {
+func newHTTPMonitor(c *config.Config, tracer connection.Tracer, bpfTelemetry *telemetry.EBPFTelemetry, offsets []manager.ConstantEditor) *usm.Monitor {
```
Comment (DanielLavie) [inline pkg/network/tracer/tracer.go:864]:
```suggestion
func newUSMMonitor(c *config.Config, tracer connection.Tracer, bpfTelemetry *telemetry.EBPFTelemetry, offsets []manager.ConstantEditor) *usm.Monitor {
```

File: pkg/network/tracer/tracer.go [hunk 2]
```diff
@@ -51,7 +52,7 @@ type Tracer struct {
 	state        network.State
 	conntracker  netlink.Conntracker
 	reverseDNS   dns.ReverseDNS
-	httpMonitor  *http.Monitor
+	httpMonitor  *usm.Monitor
```
Comment (DanielLavie) [inline pkg/network/tracer/tracer.go:55]:
```suggestion
	usmMonitor  *usm.Monitor
```

File: pkg/network/tracer/tracer.go [hunk 3]
```diff
@@ -860,11 +861,11 @@ func (t *Tracer) DebugDumpProcessCache(ctx context.Context) (interface{}, error)
 	return nil, nil
 }
 
-func newHTTPMonitor(c *config.Config, tracer connection.Tracer, bpfTelemetry *telemetry.EBPFTelemetry, offsets []manager.ConstantEditor) *http.Monitor {
+func newHTTPMonitor(c *config.Config, tracer connection.Tracer, bpfTelemetry *telemetry.EBPFTelemetry, offsets []manager.ConstantEditor) *usm.Monitor {
 	// Shared with the HTTP program
```
Comment (DanielLavie) [inline pkg/network/tracer/tracer.go:865]:
```suggestion
	// Shared with the USM program
```

File: pkg/network/tracer/tracer_windows.go [hunk 4]
```diff
@@ -42,7 +42,7 @@ type Tracer struct {
 	stopChan        chan struct{}
 	state           network.State
 	reverseDNS      dns.ReverseDNS
-	httpMonitor     http.Monitor
+	httpMonitor     usm.Monitor
```
Comment (DanielLavie) [inline pkg/network/tracer/tracer_windows.go:45]:
```suggestion
	usmMonitor     usm.Monitor
```

File: pkg/network/tracer/tracer_windows.go [hunk 5]
```diff
@@ -279,16 +279,16 @@ func (t *Tracer) DebugDumpProcessCache(ctx context.Context) (interface{}, error)
 	return nil, ebpf.ErrNotImplemented
 }
 
-func newHttpMonitor(c *config.Config, dh driver.Handle) http.Monitor {
+func newHttpMonitor(c *config.Config, dh driver.Handle) usm.Monitor {
```
Comment (DanielLavie) [inline pkg/network/tracer/tracer_windows.go:282]:
```suggestion
func newUSMMonitor(c *config.Config, dh driver.Handle) usm.Monitor {
```

File: pkg/network/usm/kernel_requirements.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+import (
+	ddebpf "github.com/DataDog/datadog-agent/pkg/ebpf"
+	"github.com/DataDog/datadog-agent/pkg/network/config"
+	"github.com/DataDog/datadog-agent/pkg/util/kernel"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// errNotSupported indicates that the current host doesn't fullfil the
+// requirements for HTTP monitoring
```
Comment (DanielLavie) [inline pkg/network/usm/kernel_requirements.go:19]:
```suggestion
// requirements for USM monitoring
```

File: tasks/system_probe.py [hunk 7]
```diff
@@ -248,7 +248,7 @@ def ninja_runtime_compilation_files(nw):
     runtime_compiler_files = {
         "pkg/collector/corechecks/ebpf/probe/oom_kill.go": "oom-kill",
         "pkg/collector/corechecks/ebpf/probe/tcp_queue_length.go": "tcp-queue-length",
-        "pkg/network/protocols/http/compile.go": "http",
+        "pkg/network/usm/compile.go": "http",
```
Comment (DanielLavie) [inline tasks/system_probe.py:251]: Please open a Jira ticket for changing the `http` name to `USM`

**Review by DanielLavie — CHANGES_REQUESTED** [review]

**Review by hmahmood — APPROVED** [review]

File: pkg/network/tracer/tracer.go [hunk 8]
```diff
@@ -749,10 +750,10 @@ func (t *Tracer) DebugEBPFMaps(maps ...string) (string, error) {
 	if err != nil {
 		return "", err
 	}
-	if t.httpMonitor == nil {
+	if t.usmMonitor == nil {
 		return "tracer:\n" + tracerMaps, nil
 	}
-	httpMaps, err := t.httpMonitor.DumpMaps(maps...)
+	httpMaps, err := t.usmMonitor.DumpMaps(maps...)
```
Comment (DanielLavie) [inline pkg/network/tracer/tracer.go:756]:
```suggestion
	usmMaps, err := t.usmMonitor.DumpMaps(maps...)
```

**Review by DanielLavie — DISMISSED** [review]

**Review by brycekahle — APPROVED** [review]
Shallow review on just `tasks/system_probe.py`

**Review by richl-dd — APPROVED** [review]
Looks good for WKIT files and the name updates.

File: pkg/network/protocols/http/driver_interface.go [hunk 9]
```diff
@@ -46,12 +46,12 @@ type WinHttpTransaction struct {
...
 	// ContentLength uint32
 }
-type httpDriverInterface struct {
+type HttpDriverInterface struct {
 	driverHTTPHandle  driver.Handle
 	driverEventHandle windows.Handle
 
 	readMux     sync.Mutex
-	dataChannel chan []WinHttpTransaction
+	DataChannel chan []WinHttpTransaction
```
Comment (guyarb) [inline pkg/network/protocols/http/driver_interface.go:54]: why did we make it exported?

File: pkg/network/usm/kernel_requirements.go [hunk 10]
```diff
@@ -0,0 +1,40 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
```
Comment (guyarb) [inline pkg/network/usm/kernel_requirements.go:1]: the file name is not relevant to its content
consider renaming it

File: pkg/network/usm/monitor.go [hunk 11]
```diff
@@ -49,10 +50,10 @@ type Monitor struct {
 	httpConsumer    *events.Consumer
 	http2Consumer   *events.Consumer
 	ebpfProgram     *ebpfProgram
-	httpTelemetry   *telemetry
-	http2Telemetry  *telemetry
-	statkeeper      *httpStatKeeper
-	http2Statkeeper *httpStatKeeper
+	httpTelemetry   *http.Telemetry
+	http2Telemetry  *http.Telemetry
+	statkeeper      *http.HttpStatKeeper
```
Comment (guyarb) [inline pkg/network/usm/monitor.go:55]: consider changing it to `httpStatKeeper` to have a similar convention to `http2StatKeeper`

**Review by guyarb — COMMENTED** [review]
LGTM
left couple of suggestions

File: pkg/network/tracer/tracer.go [hunk 12]
```diff
@@ -860,11 +861,11 @@ func (t *Tracer) DebugDumpProcessCache(ctx context.Context) (interface{}, error)
 	return nil, nil
 }
 
-func newHTTPMonitor(c *config.Config, tracer connection.Tracer, bpfTelemetry *telemetry.EBPFTelemetry, offsets []manager.ConstantEditor) *http.Monitor {
-	// Shared with the HTTP program
+func newUSMMonitor(c *config.Config, tracer connection.Tracer, bpfTelemetry *telemetry.EBPFTelemetry, offsets []manager.ConstantEditor) *usm.Monitor {
```
[the remark this answers is not in the document]
Comment (DanielLavie) [reply pkg/network/tracer/tracer.go:864]: Add Jira task

File: pkg/network/usm/kernel_requirements.go (hunk 10 above)
Comment (DanielLavie) [reply pkg/network/usm/kernel_requirements.go:1]: Renamed

**Review by guyarb — APPROVED** [review]

**Review by guyarb — CHANGES_REQUESTED** [review]
The changes broke tests in shared_libraries

while you moved  `shared_libraries_test.go`, you didn't move `pkg/network/protocols/http/testutil/sowatcher_client/sowatcher_client.go` so in the CI we're not able to find it and tests are failing

review #16650 to see what changes do you need to do

File: pkg/network/tracer/tracer.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (DanielLavie) [reply pkg/network/tracer/tracer.go:864]: https://datadoghq.atlassian.net/browse/USMO-362

**Review by guyarb — APPROVED** [review]

**Pull request merged**